src/compiler/checker.ts

```diff
@@ -742,6 +743,7 @@ namespace ts {
         const stringNumberSymbolType = getUnionType([stringType, numberType, esSymbolType]);
         const keyofConstraintType = keyofStringsOnly ? stringType : stringNumberSymbolType;
         const numberOrBigIntType = getUnionType([numberType, bigintType]);
+        const templateConstraintType = getUnionType([stringType, numberType, booleanType, bigintType]);

```

Comment:
It is kind of funny that we allow template strings to concatenate with literally any other type. Probably a good opportunity for a pedantic flag.